### DISCUSSION TITLE
Support of Common Expressions in Friction's Expression Editor

### DISCUSSION THREAD

**Discussion by etakanaka** [friction2d/friction#242] [Ideas]
Hi Friction Team,

First off, thank you for developing such an exciting open-source tool! The **Powerful Expression Editor** is a great feature, and I appreciate the inclusion of ECMAScript, syntax highlighting, and autocompletion.

As someone coming from Adobe After Effects, I often rely on certain pre-built expressions for automating animations. Some common expressions that I (and many other users) find extremely useful include:

1. **Wiggle()** – Randomly varies values like position, rotation, or scale.
2. **Time** – Drives animation based on time, great for creating dynamic effects.
3. **Loop Expressions** – To create repeating animations without duplicating keyframes (e.g., `loopOut("cycle")` and `loopIn("pingpong")`).
4. **ValueAtTime()** – Retrieves a property value at a specific time.
5. **Linear() and Ease()** – For remapping ranges of values with and without easing.
6. **Math.random()** – For generating random values.
7. **Clamp()** – Restricting a property within a specified range.

I would love to see these expressions or equivalent functionality natively supported in Friction's expression editor. It would make transitioning to Friction smoother for users familiar with tools like After Effects. Additionally, features like autocompletion of these common expressions would be incredibly useful for enhancing workflow.

If these expressions (or their equivalents) are already supported in Friction, could you kindly provide some instructions or examples on how to implement them using the ECMAScript expression editor? Any guidance on achieving similar automation or parameter control would be greatly appreciated!

Thanks again for your hard work on Friction, and I look forward to exploring its full potential!

Best regards,

**Comment by rodlie** [comment]
I can add a preset feature for 'Definitions' in the expression editor, that way users can make their own helper functions and load them as needed (and share them).

> Clamp() – Restricting a property within a specified range.

Example for clamp:
![Screenshot from 2024-09-07 19-51-00](https://github.com/user-attachments/assets/e3d8f697-ee24-4975-8741-4e274f7cad8d)

> Math.random() – For generating random values.

Most `Math` functions are supported.
![friction-math](https://github.com/user-attachments/assets/639ea6cb-7ed8-41d9-b61a-61763f0611df)

> Linear() and Ease() – For remapping ranges of values with and without easing.

We have functions for easing, but they are not exposed to the user, this can be fixed.

**Reply by rodlie** [reply]
Issue to track this feature: https://github.com/friction2d/friction/issues/244

**Reply by etakanaka** [reply]
That would be very helpful, thank you, much appreciated

**Comment by pgilfernandez** [comment]
When you start exploring the limits of Friction, you arrive to expressions and realize how powerful they are, even more, as they are translated to "transforms" when exporting to SVG, they get even more powerful.

I'm trying to build my little functions library and this feature would be not only great for those who want to quickly access their common functions but also for newcomers that need to learn and understand expressions.

BTW, here are the ones I tried and are working (I guess they could be improved but they are my very first ones):

## Wiggle
```
// DEFINITIONS
function generateNoise(t, scale = 1, strength = 1, phase = 0, detail = 1) {
    // Fixed permutation for consistency
    const permutation = Array.from({ length: 256 }, (_, i) => i).concat(Array.from({ length: 256 }, (_, i) => i));
    
    function fade(t) {
        return t * t * t * (t * (t * 6 - 15) + 10);
    }

    function lerp(a, b, t) {
        return a + t * (b - a);
    }

    function grad(hash, x) {
        const h = hash & 15;
        const gradient = 1 + (h & 7); // Gradient range between 1 and 8
        return (h & 8 ? -gradient : gradient) * x;
    }

    function perlinNoise(x) {
        const X = Math.floor(x) & 255;
        x -= Math.floor(x);
        const u = fade(x);

        const a = permutation[X];
        const b = permutation[X + 1];

        return lerp(grad(a, x), grad(b, x - 1), u);
    }

    let amplitude = strength;
    let frequency = scale;
    let total = 0;

    for (let i = 0; i < detail; i++) {
        total += perlinNoise((t + phase) * frequency) * amplitude;
        frequency *= 2;       // Double frequency for each octave
        amplitude /= 2;       // Halve amplitude for each octave
    }
    return total;
}
```
```
// BINDINGS
frame = $frame;
value = $value; // in case you want to export to SVG remove this line and create a constant in "calculate" with your desired value (if not Friction will crash)
```
```
// CALCULATE
const fps = 30;
const time = frame / fps;
//const value = 200; // uncomment if you are exporting to SVG
//                      generateNoise(time, scale, strength, phase, detail)
const noise = generateNoise(time, 0.6, 20, 0, 4);
return noise+value
```

## Loops

It is possible to create **infinite loops** by mixing some Friction features (it doesn't work for SVG export though):

1) Create a new scene and set up as desired. It will be a container of just the looped animation.
2) Add the assets and animate them
3) Then switch to the main scene, click on the canvas a select `link scene` choosing the scene that contains the animation you want to loop.
4) select the linked scene object, open the contextual menu (clicking the secondary mouse button) and turn on `Frame remapping`. It will activate a new parameter inside the linked scene called "**Frame**".
5) Click on the Frame parameter with the secondary button and choose `Set Expression`
6) Under **Bindings**:
```frame = $frame```

7) Under **Calculate**
```return frame%30```
where "30" is just an example but it is the last frame you want to loop, that is, if your loop is within frame number 0 and frame number 30 then set that number in **Calculate** field.

## Orbit
Here you need to create several "custom properties" to have more control over the script:
```
// CUSTOM PROPERTIES
centerX (pixels)
centerY (pixels)
radius (pixels)
speed (frames)
direction (1 clockwise, -1 counter clockwise)
offset (degrees)
```
```
// BINDINGS
centerX = properties.centerX;
centerY = properties.centerY;
radius = properties.radius;
speed = properties.speed;
direction = properties.direction;
offset = properties.offset;
frame = $frame;
```
```
// CALCULATE (under translate X)
var offset = offset * (Math.PI / 180);
var centerOfComp = [centerX, centerY];
var circleOverTime = [ Math.sin( (frame/speed) * 2 * Math.PI * direction + offset ) * radius, -Math.cos( (frame/speed)* 2 * Math.PI * direction + offset ) * radius ];  
return centerOfComp[0] + circleOverTime[0];
```
```
// CALCULATE (under translate Y)
var offset = offset * (Math.PI / 180);
var centerOfComp = [centerX, centerY];
var circleOverTime = [ Math.sin( (frame/speed) * 2 * Math.PI * direction + offset ) * radius, -Math.cos( (frame/speed) * 2 * Math.PI * direction + offset ) * radius ];  
return centerOfComp[1] + circleOverTime[1];
```

EDIT: updated Orbit expressions with more control
EDIT2: fixed Loop chapter with a better technique (it not only involve Expressions but it's a definitive one)

**Comment by zaynopenapp** [comment]
Yes. We need valueAttime(). How to get property of value of points of path??(Expression)

. Like in synfig i can get value of point in bline. valueAttime() i used convert timeloop set duration zero(0). Seek value by set Link time.

**Reply by pgilfernandez** [reply]
> Yes. We need valueAttime(). How to get property of value of points of path??(Expression)
> 
> . Like in synfig i can get value of point in bline. valueAttime() i used convert timeloop set duration zero(0). Seek value by set Link time.

Yeah, I've used "value of points of path" in Synfig and I agree it would be nice to have in Expressions. I haven't tried but I bet there is some kind of function that let us get the position parameters of the "complete" parameter applied to a path...

**Reply by pgilfernandez** [reply]
> How to get property of value of points of path??(Expression)

I've been tinkering with the idea and got something that works out of the box, that is a workaround:

1. isolate/locate the path you want to use and get its code, something like this:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Friction - https://friction.graphics -->

<svg viewBox="0 0 1920 1080" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
 <g transform="translate(0 0)">
  <g transform="translate(121.668 139.14)">
   <g transform="rotate(0)">
    <g transform="scale(1 1)">
     <g transform="skewX(0) skewY(0)">
      <g transform="translate(0 0)" opacity="1">
       <path d="M0 0C524.965 17.3065 346.131 398.051 623.036 565.347C899.941 732.644 1381.64 449.97 1214.34 124.03C1047.05 -201.91 623.036 23.0754 686.493 314.402C749.95 605.729 1468.17 969.167 1667.2 297.096" stroke-linecap="round" stroke-width="10" stroke-linejoin="round" stroke="rgba(0, 102, 255, 1)" fill="none" id="path2"/>
      </g>
     </g>
    </g>
   </g>
  </g>
 </g>
 <defs/>
</svg>
```
Ask your preferred AI to turn it into a bezier matrix:
```
var path = [
    [[0.0, 0.0], [524.965, 17.3065], [346.131, 398.051], [623.036, 565.347]],
    [[623.036, 565.347], [899.941, 732.644], [1381.64, 449.97], [1214.34, 124.03]],
    [[1214.34, 124.03], [1047.05, -201.91], [623.036, 23.0754], [686.493, 314.402]],
    [[686.493, 314.402], [749.95, 605.729], [1468.17, 969.167], [1667.2, 297.096]]
];
```
2. Now introduce it in `X` and `Y` as expressions with some more code:
```
// BINDINGS
complete = properties.complete;
frame = $frame;
offsetX = path.transform.translation.x;
offsetY = path.transform.translation.y;
```
```
// CALCULATE (scripts)
var path = [
    [[0.0, 0.0], [524.965, 17.3065], [346.131, 398.051], [623.036, 565.347]],
    [[623.036, 565.347], [899.941, 732.644], [1381.64, 449.97], [1214.34, 124.03]],
    [[1214.34, 124.03], [1047.05, -201.91], [623.036, 23.0754], [686.493, 314.402]],
    [[686.493, 314.402], [749.95, 605.729], [1468.17, 969.167], [1667.2, 297.096]]
];

var numSegments = path.length;

var t_global = clamp(complete * numSegments, 0, numSegments);  

var segmentIndex = Math.floor(t_global);
segmentIndex = clamp(segmentIndex, 0, numSegments - 1);

var t_local = t_global - segmentIndex;

var P0 = path[segmentIndex][0];
var P1 = path[segmentIndex][1];
var P2 = path[segmentIndex][2];
var P3 = path[segmentIndex][3];

var x = Math.pow(1 - t_local, 3) * P0[0] +
        3 * Math.pow(1 - t_local, 2) * t_local * P1[0] +
        3 * (1 - t_local) * Math.pow(t_local, 2) * P2[0] +
        Math.pow(t_local, 3) * P3[0];

var y = Math.pow(1 - t_local, 3) * P0[1] +
        3 * Math.pow(1 - t_local, 2) * t_local * P1[1] +
        3 * (1 - t_local) * Math.pow(t_local, 2) * P2[1] +
        Math.pow(t_local, 3) * P3[1];

return x + offsetX;
```

and do the same to `Y` value and change the final `return x + offsetX` with `return y + offsetY`.
You may need to tweak some more things but as it is long to explain it's easier for me to just share with you the project I played with:

![expression_point_of_path_v04b](https://github.com/user-attachments/assets/ffe76518-5272-4802-a880-ce29a11027ae)

[expression_point_of_path_v04.zip](https://github.com/user-attachments/files/18905643/expression_point_of_path_v04.zip)

Let me know 😉
Cheers

PS - it could look to much code to copy every time but take into account that there is an upcoming Friction v1.0 RC1 that adds support for Expressions Presets that are super cool and handy to save/import/export them 😄

<img width="900" alt="screenshot" src="https://github.com/user-attachments/assets/8e381ba4-3b40-4992-a108-44b6cef476b2" />

**Reply by zaynopenapp** [reply]
Thank you. So much..

**Comment by pgilfernandez** [comment]
hi @etakanaka,

> 4\. **ValueAtTime()** – Retrieves a property value at a specific time.

The following PR adds support for this feature: https://github.com/friction2d/friction/pull/803
You'll be able to try the test builds 💪🏻
